rust/xaynet-server/src/services/messages/multipart.rs

```diff
@@ -0,0 +1,456 @@
+        }
+        self.data.insert(id, data);
+    }
+
+    /// Aggregate all the chunks. This method should only be called
+    /// when all the chunks are here, otherwise the aggregated message
+    /// will be invalid.
+    fn into_message(self) -> Result<Message, DecodeError> {
+        // FIXME: this is highly inefficient. We end up allocating a
+        // potentially huge vector (dozens of MB is worst but realistic
+        // cases). In theory we could parse the message directly from the
+        // BTreeMap. But that would require parsing from a stream of bytes
+        // instead of an array, which xaynet_core doesn't support atm.
+        let bytes = self
+            .data
+            .into_iter()
+            .map(|(_, chunk)| chunk)
+            .fold(vec![], |mut acc, chunk| {
+                acc.extend(chunk);
+                acc
+            });
+        let payload = match self.tag {
+            Tag::Sum => Sum::from_bytes(&&bytes[..]).map(Into::into)?,
+            Tag::Update => Update::from_bytes(&&bytes[..]).map(Into::into)?,
+            Tag::Sum2 => Sum2::from_bytes(&&bytes[..]).map(Into::into)?,
+        };
+        let message = Message {
+            signature: None,
+            participant_pk: self.participant_pk,
+            coordinator_pk: self.coordinator_pk,
+            tag: self.tag,
+            is_multipart: false,
+            payload,
+        };
+        Ok(message)
+    }
+}
+
+#[derive(Debug, Hash, Eq, PartialEq, Clone)]
+pub struct MessageId {
+    message_id: u16,
+    participant_pk: PublicSigningKey,
+}
+
+pub struct MultipartHandler {
+    messages: HashMap<MessageId, MultipartMessage>,
+}
+
+impl MultipartHandler {
+    #[allow(dead_code)]
+    pub fn new() -> Self {
+        Self {
+            messages: HashMap::new(),
+        }
+    }
+}
+
+impl Service<Message> for MultipartHandler {
+    type Response = Option<Message>;
+    type Error = ServiceError;
+    type Future = Ready<Result<Self::Response, Self::Error>>;
+
+    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, message: Message) -> Self::Future {
+        if !message.is_multipart {
+            return ready_ok(Some(message));
+        }
+
+        if let Message {
+            tag,
+            participant_pk,
+            coordinator_pk,
+            payload: Payload::Chunk(chunk),
+            ..
+        } = message
+        {
+            let id = MessageId {
+                message_id: chunk.message_id,
+                participant_pk,
+            };
+            let mp_message = self
+                .messages
+                .entry(id.clone())
+                .or_insert_with(|| MultipartMessage::new(tag, participant_pk, coordinator_pk));
+            mp_message.add_chunk(chunk);
+            if mp_message.has_all_chunks() {
+                // We just upserted that entry so we can safely unwrap
```

Comment:
I don't really understand that comment here

Reply:
It was unclear indeed. I reformulated.